### DISCUSSION TITLE
Multiple 'REL_OP' options?

### DISCUSSION THREAD

**Discussion by ninjalu** [explosion/spaCy#10283] [Help: Coding & Implementations]
I am using dependency matcher to match close questions, like
Are you....? etc.

For it to work, I need to make sure that 'are' is a parent of 'you' and also precedes 'you'.

At the moment I can use one 'LEFT_ID' and 'RIGHT_ID' pair to define the 'parenthood', as follows:

```
        {
            'LEFT_ID': 'anchor_aux',
            'REL_OP': '>',
            'RIGHT_ID': 'nsubj',
            'RIGHT_ATTRS': {
                'DEP': 'nsubj'
            }
        },

```
And another pair to define the order, as follows:
```
        {
            'LEFT_ID': 'anchor_aux',
            'REL_OP': '.*',
            'RIGHT_ID': 'nsubj2',
            'RIGHT_ATTRS': {
                'DEP': 'nsubj'
            }
        },
```
However this will become a problem if somewhere else in the sentence there is another `nsubj`. 

Is it possible to define the relationships between these two nodes BOTH '>' and '.*' at the same time?

**Comment by polm** [comment]
You can't pass multiple REL_OPs to the DependencyParser for a single token. 

In this case I think `.*` does what you want? It's true it doesn't define which token is the parent of the other, but given your left token is a verb and your right token is `nsubj` I don't think those relations can be reversed meaningfully. If you define the verb as being the ROOT reversal is also impossible.

Another thing is that if you are actually looking just for `Are you` type patterns, honestly maybe you don't need the dependency matcher? I can imagine variations ("Are all of you"), but it seems like just matching the literal tokens will probably cover the vast majority of cases - you might want to try that if you haven't.

**Reply by ninjalu** [reply]
Thanks very much for your quick reply. 
The reason I would want to retain `>` is to differentiate from `>>`, so when a `nsubj` appears in a clause in the sentence, it wouldn't become a problem. For instance:
`I am from Paris, where people enjoy cheese and wine.`
The root `are` is before `nsubj` people, so with only `.*` you would detect a false positive question. 

I probably haven't explained clearly: the aim is to detect close questions in general, so `Are all of you` needs to be detected as well.

**Reply by polm** [reply]
I see, I guess you can't exclude cases like that using just the operators. 

Generally what you can do in this case is create a rule that includes false positives and filter them afterwards. For example in your cheese and wine sentence, you can take the nsubj and go up the tree and discard it if the verb it depends on is not the one you are looking for. 

One detail: can you clarify what you mean by "close question"? Do you mean a [closed-ended question](https://en.wikipedia.org/wiki/Closed-ended_question) or something else? 

One thing about questions is it might just be easier to check for question marks than verb inversion.

**Reply by ninjalu** [reply]
Thanks for your reply! 

Yes, by "close question", I meant "close-ended" questions.

Checking for questions marks work sometimes, but with transcripts that I am working on, the data can be messy, and things like punctuations are not always followed to the best practice.

And also because human speech is never clean and textbook-like, repetitions, gramma mistakes adds another layer to the complexity. Hence exceptions are not working very well (because of too many edge cases). In the past days of testing and writing up exceptions, I am more and more convinced that multiple "REL_OP" is the only reliable way to resolve this problem.

I wonder if adding the feature of multiple "REL_OP" is way too complicated, or perhaps somebody like me without a lot of knowledge of spaCy source code can give it a go.

**Reply by polm** [reply]
Multiple REL_OP is not easy to add because the data is just designed to have one REL_OP per dictionary. You could add multiple dictionaries with the same RIGHT_ID more easily; currently that causes an error because we check for it, but maybe if you removed the error check you could get an intersection of the constraints. I have not tested that.

> I am more and more convinced that multiple "REL_OP" is the only reliable way to resolve this problem.

Maybe you are talking about other cases but since all `.*` is doing is literally checking the token indices it seems trivial to just post-process things matched with the `>` rule? Something like...

```
for match in matches:
    match_id, toks = match
    aux = doc[toks[0]]
    nsubj = doc[toks[1]]
    if aux.i > nsubj.i: continue
    # do the actual processing here
```

**Reply by ninjalu** [reply]
> currently that causes an error because we check for it, but maybe if you removed the error check you could get an intersection of the constraints. I have not tested that.

Thanks. I tried that today, and it resulted in the dependency matcher looking for new nodes anyway, not the nodes previous defined, despite the ID being the same. So effectively, removing the warning didn't change the result. 

Could you guide us what's the next step I could do in terms of editing source code?

> Maybe you are talking about other cases but since all .* is doing is literally checking the token indices it seems trivial to just post-process things matched with the > rule? Something like...

```
for match in matches:
    match_id, toks = match
    aux = doc[toks[0]]
    nsubj = doc[toks[1]]
    if aux.i > nsubj.i: continue
    # do the actual processing here
```

This could work, in this case. But I would prefer the first solution, if I can help it! it's a lot cleaner and easier to use for other patterns

Many thanks!

**Reply by polm** [reply]
> Could you guide us what's the next step I could do in terms of editing source code?

Sorry, I haven't really thought about it before. Maybe the easiest thing would be adding a new REL_OP?

**Reply by ninjalu** [reply]
> Maybe the easiest thing would be adding a new REL_OP?

That's not a bad idea! Thanks I will read the source code and try to figure out how to do that.

**Reply by ninjalu** [reply]
I have added four additional "REL_OP": `>.*` for right child, `>;*` left child, `<.*` right parent, and `<;*` left parent to `dependencymatcher.pyx`. It seems to work as a treat for me,
Would you consider merging the changes, if I push it up? I could have a local version of spaCy, but I don't want to miss out on future improvement from your side. 

Let me know what's the best way forward. Many thanks!

the changes are, in `DependencyMatcher`:

```
        self._ops = {
            "<": self._dep,
            ">": self._gov,
            "<<": self._dep_chain,
            ">>": self._gov_chain,
            ".": self._imm_precede,
            ".*": self._precede,
            ";": self._imm_follow,
            ";*": self._follow,
            "$+": self._imm_right_sib,
            "$-": self._imm_left_sib,
            "$++": self._right_sib,
            "$--": self._left_sib,
# New REL_OP as below:
            ">.*": self._right_child,
            ">;*": self._left_child,
            "<.*": self._right_parent,
            "<;*": self._left_parent,
        }
```

And additional of these

```
    def _right_child(self, doc, node):
        return [doc[child.i] for child in doc[node].children if child.i > node]
    
    def _left_child(self, doc, node):
        return [doc[child.i] for child in doc[node].children if child.i < node]

    def _right_parent(self, doc, node):
        if doc[node].head == doc[node]:
            return []
        elif doc[node].head.i > node:
            return [doc[node].head]
        return []
    
    def _left_parent(self, doc, node):
        if doc[node].head == doc[node]:
            return []
        elif doc[node].head.i < node:
            return [doc[node].head]
        return []
```

**Reply by polm** [reply]
Oh wow, thanks! That actually does look pretty simple. 

If you could go ahead an open a PR that'd make it easy for the whole team to take a look at it.

**Reply by ninjalu** [reply]
Thanks a lot! I have created PR #10371 with the above changes.

Never attempted to contribute to open source before! It's a very small change, and not sure whether it will get merged. Still I am buzzing :D

**Reply by polm** [reply]
Thanks for making the PR, and congrats on your first! The team will take a look at it and get back to you.
